applications/new_schema.py

```diff
@@ -0,0 +1,75 @@
+import django_filters
+import graphene
+from graphene_django.filter import DjangoFilterConnectionField
+from graphene_django.types import DjangoObjectType
+from graphql_jwt.decorators import login_required, superuser_required
+
+from .models import BerthApplication, BerthSwitch, HarborChoice
+from .schema import BerthSwitchType as OldBerthSwitchType
+
+
+class HarborChoiceType(DjangoObjectType):
+    harbor = graphene.String()
+
+    class Meta:
+        model = HarborChoice
+        exclude = ("id", "application")
+
+    def resolve_harbor(self, info, **kwargs):
+        if self.harbor:
+            return self.harbor.servicemap_id
+        return None
+
+
+class BerthSwitchType(OldBerthSwitchType):
+    harbor = graphene.String()
+
+    class Meta:
+        model = BerthSwitch
+
+    def resolve_harbor(self, info, **kwargs):
+        if self.harbor:
+            return self.harbor.servicemap_id
+        return None
+
+
+class BerthApplicationFilter(django_filters.FilterSet):
+    has_berth_switch = django_filters.BooleanFilter(
+        field_name="berth_switch", method="filter_berth_switch"
+    )
+
+    def filter_berth_switch(self, queryset, name, value):
+        lookup = "__".join([name, "isnull"])
+        return queryset.filter(**{lookup: not value})
+
+
+class BerthApplicationTypeNode(DjangoObjectType):
+    boat_type = graphene.String()
+    harborchoice_set = graphene.List(HarborChoiceType)
+
+    class Meta:
+        model = BerthApplication
+        interfaces = (graphene.relay.Node,)
+        exclude = ("chosen_harbors",)
+
+    def resolve_boat_type(self, info, **kwargs):
+        if self.boat_type:
+            return self.boat_type.id
+        return None
+
+    def resolve_harborchoice_set(self, info, **kwargs):
+        if self.harborchoice_set.count():
+            return self.harborchoice_set.all()
+        return None
+
+
+class Query:
+    application = graphene.relay.Node.Field(BerthApplicationTypeNode)
+    applications = DjangoFilterConnectionField(
+        BerthApplicationTypeNode, filterset_class=BerthApplicationFilter
+    )
+
+    @login_required
+    @superuser_required
+    def resolve_applications(self, info, **kwargs):
+        return BerthApplication.objects.all()
```

Comment:
you could do some `select_related()` and `prefetch_related()` to optimize the query somewhat. check [from here onwards](https://github.com/City-of-Helsinki/berth-reservations/blob/develop/resources/schema.py#L612) for some references

Reply:
What should be prefetched? I'm thinking `chosen_harbors` and `boat_type` at least (since they're the only ones with foreign keys), but what else should be included?

Reply:
yes, `boat_type`, `berth_switch` and `chosen_harbors` (which would in practice mean `harborchoice_set` and `harborchoice_set__harbor` or something) 🤔

Reply:
and `berth_switch` recursively, meaning `berth_switch__harbor` and `berth_switch__reason`

Reply:
and `berth_switch__reason__translations` 😅

Reply:
👍

Reply:
So,

```python
return BerthApplication.objects.prefetch_related(
    "boat_type",
    "berth_switch",
    "berth_switch__harbor",
    "berth_switch__reason",
    "berth_switch__reason__translations",
    "harborchoice_set__harbor",
)
```

Reply:
so, `prefetch_related` is for m2m relations. `select_related` is for FKs:
```python
return BerthApplication.objects.select_related(
    "boat_type",
    "berth_switch",
    "berth_switch__harbor",
    "berth_switch__reason",
    "harborchoice_set__harbor",
).prefetch_related("berth_switch__reason__translations")
```

Reply:
I was just reading about that. 😅